File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/scan/cardcapture/BusinessCardPostProcessor.kt
/*
 * BusinessCardPostProcessor.kt
 *
 * Hand-off point between the Business Card capture surface and
 * the existing OCR + contact-extraction pipeline. Given a
 * captured Bitmap (full-resolution still from CameraX) and the
 * most recent valid quad in still-image coordinates, this
 * post-processor:
 *
 *   1. Perspective-corrects the source corners to (0,0)→
 *      (1011,0)→(1011,637)→(0,637) using Android's
 *      `Matrix.setPolyToPoly` + Canvas (GPU-backed on most
 *      devices; faster than the per-pixel Kotlin warp in
 *      CardImageOps).
 *   2. Saves the result as a JPEG inside AttachmentStorage.
 *   3. Builds a [DocumentScanResult] with the warped JPEG as
 *      the single page + preview, and routes it through
 *      [ScanFlowController.onScanComplete] with
 *      `category = "Business Card"` so the existing scan-detail
 *      screen picks up the BusinessCardExtractor flow.
 *
 * Manual capture (user tapped the shutter without a valid
 * stability lock): callers pass `quadInBitmap = null` and we
 * fall back to the guide rect as the "card quad" — the user
 * was free-handing it, so a center-of-frame heuristic is the
 * best we can do without rejecting their tap.
 *
 * The CardImageOps perspective-warp implementation lives in
 * pure Kotlin so it's unit-testable on the JVM; the Android
 * Matrix path here is the production code path that runs on
 * the GPU.
 */

package app.quickink.mobile.features.scan.cardcapture

import android.content.Context
import android.graphics.Bitmap
import android.graphics.Canvas
import android.graphics.Matrix
import android.graphics.Paint
import android.net.Uri
import app.quickink.mobile.features.scan.ScanFlowController
import app.releaf.mobile.data.common.AttachmentStorage
import app.releaf.mobile.data.common.Uuidv7
import app.releaf.shared.scan.DocumentScanResult
import java.io.File

object BusinessCardPostProcessor {

    /** Warped output size — spec callout. */
    const val OUTPUT_WIDTH  = 1012
    const val OUTPUT_HEIGHT = 638

    /**
     * Run the full post-process and hand the resulting capture
     * to the controller. Returns the warped JPEG's file:// URI
     * so the caller (the surface) can show an in-screen confirm
     * animation if it wants.
     *
     * The user-visible region of the bitmap is the FILL_CENTER
     * center crop whose aspect matches [viewWidth]:[viewHeight]
     * (the on-screen canvas behind the overlay). The guide rect
     * we crop to is the 70%-of-width / 1.586:1 / 45%-vertical
     * sub-rect inside THAT visible region — identical math to
     * the overlay's draw + the detector's IoU check, so what
     * gets warped is exactly what the user framed.
     *
     * Threading: the bitmap warp + JPEG encode are CPU-bound
     * (Canvas + Bitmap.compress); the controller's
     * `onScanComplete` kicks coroutines internally so it's
     * non-blocking. Call from a background thread.
     */
    fun process(
        context: Context,
        source: Bitmap,
        quadInBitmap: DetectedQuad?,
        viewWidth: Float,
        viewHeight: Float,
        controller: ScanFlowController,
    ): Uri? {
        val guideInBitmap = computeGuideInBitmap(source, viewWidth, viewHeight)
        val quad = quadInBitmap ?: guideInBitmap.asQuad()
        val warped = warpToCardSize(source, quad) ?: return null
        val jpegUri = saveJpeg(context, warped) ?: return null
        // Source bitmap is owned by us at this point — recycle to
        // avoid holding a multi-megapixel allocation while OCR
        // runs. The warped output is in `jpegUri`; the source's
        // pixels are no longer needed.
        if (!source.isRecycled) source.recycle()
        if (!warped.isRecycled) warped.recycle()

        // Build a DocumentScanResult that the existing controller
        // can swallow whole. pdfUri stays null — business card
        // captures don't ship a PDF artifact (the Library card
        // renders fine from the preview JPEG, and the export
        // sheet can synthesize a PDF later if needed). pageUris
        // is a single-entry list with the warped JPEG; OCR reads
        // file:// URIs the same way it reads content:// ones.
        val result = DocumentScanResult(
            pdfUri     = null,
            previewUri = jpegUri,
            pageUris   = listOf(jpegUri),
        )
        controller.onScanComplete(
            result   = result,
            category = "Business Card",
        )
        return jpegUri
    }

    /**
     * Perspective-warp [source] using [quad] as the source
     * corners and (0,0)→(1011,0)→(1011,637)→(0,637) as the
     * destination corners. Output is a fresh [Bitmap] sized to
     * [OUTPUT_WIDTH] × [OUTPUT_HEIGHT].
     */
    fun warpToCardSize(source: Bitmap, quad: DetectedQuad): Bitmap? {
        val srcPoints = floatArrayOf(
            quad.tl.x, quad.tl.y,
            quad.tr.x, quad.tr.y,
            quad.br.x, quad.br.y,
            quad.bl.x, quad.bl.y,
        )
        val dstPoints = floatArrayOf(
            0f,                                0f,
            (OUTPUT_WIDTH - 1).toFloat(),      0f,
            (OUTPUT_WIDTH - 1).toFloat(),      (OUTPUT_HEIGHT - 1).toFloat(),
            0f,                                (OUTPUT_HEIGHT - 1).toFloat(),
        )
        val matrix = Matrix()
        // 4-point poly → poly. Android computes the homography
        // internally — same arithmetic as CardImageOps.getPerspectiveTransform
        // but routed through Skia's GPU path on the subsequent
        // Canvas draw.
        if (!matrix.setPolyToPoly(srcPoints, 0, dstPoints, 0, 4)) return null

        val out = Bitmap.createBitmap(
            OUTPUT_WIDTH,
            OUTPUT_HEIGHT,
            Bitmap.Config.ARGB_8888,
        )
        val canvas = Canvas(out)
        val paint = Paint().apply {
            isFilterBitmap = true     // bilinear sampling
            isAntiAlias    = true
            isDither       = true
        }
        canvas.drawBitmap(source, matrix, paint)
        return out
    }

    /**
     * Compute the in-bitmap guide rect that corresponds to the
     * on-screen overlay, accounting for the FILL_CENTER center
     * crop the PreviewView applies. Same arithmetic as the
     * detector's analyzer-guide computation; centralized in
     * [CardImageOps] so both sides agree.
     */
    fun computeGuideInBitmap(source: Bitmap, viewWidth: Float, viewHeight: Float): GuideRect {
        val visible = CardImageOps.visibleRectForViewAspect(
            imageWidth  = source.width,
            imageHeight = source.height,
            viewWidth   = viewWidth,
            viewHeight  = viewHeight,
        )
        return CardImageOps.guideRectInside(visible)
    }

    /**
     * Save [bitmap] as JPEG into [AttachmentStorage.directory] and
     * return the file:// URI. Quality 92 — high enough that the
     * text edges in the warped card stay crisp for ML Kit's
     * Latin-script text recognizer, low enough that a typical
     * card lands at ~150 KB on disk.
     */
    fun saveJpeg(context: Context, bitmap: Bitmap): Uri? {
        val dest = File(
            AttachmentStorage.directory(context),
            "${Uuidv7.generate()}.jpg",
        )
        return runCatching {
            dest.outputStream().use { out ->
                if (!bitmap.compress(Bitmap.CompressFormat.JPEG, 92, out)) return null
            }
            Uri.fromFile(dest)
        }.getOrNull()
    }
}

File: apps/quickink/ios/QuickInk/Scan/CardCapture/BusinessCardPostProcessor.swift
/*
 * BusinessCardPostProcessor.swift
 *
 * Hand-off point between the iOS Business Card capture
 * surface and the existing OCR + contact-extraction
 * pipeline. Given a captured CGImage + the most recent
 * valid quad in image coordinates, this post-processor:
 *
 *   1. Perspective-corrects the source corners to
 *      (0,0)→(1011,0)→(1011,637)→(0,637) using Core Image's
 *      `CIPerspectiveCorrection` filter (GPU-backed on all
 *      modern A-chip devices).
 *   2. Saves the result as a JPEG inside the AttachmentStorage
 *      directory.
 *   3. Calls `ScanFlowController.onScanComplete` with
 *      `category: "business-card"` so the existing scan-detail
 *      screen picks up the `BusinessCardExtractor` flow.
 *
 * Manual capture (user tapped the shutter without a valid
 * stability lock): callers pass `quad = nil` and we fall back
 * to the guide rect's corners as the "card quad" — the user
 * was free-handing it, so a center-of-frame heuristic is the
 * best we can do without rejecting their tap.
 *
 * Mirror of Android `BusinessCardPostProcessor.kt`.
 */

import Foundation
import CoreImage
import CoreGraphics
import UIKit
import ReleafCoreData
import ReleafCoreScan

public enum BusinessCardPostProcessor {

    public static let outputWidth: CGFloat  = 1012
    public static let outputHeight: CGFloat = 638

    /// Run the full post-process and hand the resulting capture
    /// to the controller. Returns the warped JPEG's file URL so
    /// the surface can show an in-screen confirm animation if
    /// it wants.
    ///
    /// The user-visible region of the image is the
    /// resizeAspectFill center crop whose aspect matches
    /// `viewWidth`:`viewHeight` (the on-screen canvas behind
    /// the overlay). The guide rect we crop to is the
    /// 70%-of-width / 1.586:1 / 45%-vertical sub-rect inside
    /// THAT visible region — identical math to the overlay's
    /// draw + the detector's IoU check, so what gets warped is
    /// exactly what the user framed.
    ///
    /// Threading: warp + JPEG encode are CPU/GPU-bound; the
    /// controller's `onScanComplete` kicks Tasks internally so
    /// it doesn't block. Call from a non-MainActor context.
    @MainActor
    public static func process(
        source: CGImage,
        quadInImage: DetectedQuad?,
        viewWidth: Float,
        viewHeight: Float,
        controller: ScanFlowController,
    ) async -> URL? {
        let guideInImage = computeGuideInImage(
            source:     source,
            viewWidth:  viewWidth,
            viewHeight: viewHeight,
        )
        let quad = quadInImage ?? guideInImage.asQuad
        guard let warpedURL = await warpAndSave(source: source, quad: quad) else {
            return nil
        }
        controller.onScanComplete(
            pdfURL:     nil,
            previewURL: warpedURL,
            pageURLs:   [warpedURL],
            category:   "business-card",
            paperSize:  .card,
        )
        return warpedURL
    }

    /// Compute the in-image guide rect that corresponds to the
    /// on-screen overlay, accounting for the resizeAspectFill
    /// center crop the preview layer applies. Centralized in
    /// [CardImageOps] so the detector + post-processor agree on
    /// the user-visible region.
    public static func computeGuideInImage(
        source: CGImage,
        viewWidth: Float,
        viewHeight: Float,
    ) -> GuideRect {
        let visible = CardImageOps.visibleRectForViewAspect(
            imageWidth:  source.width,
            imageHeight: source.height,
            viewWidth:   viewWidth,
            viewHeight:  viewHeight,
        )
        return CardImageOps.guideRectInside(visible)
    }

    /// Apply Core Image's `CIPerspectiveCorrection` filter to
    /// map [quad] → axis-aligned `outputWidth × outputHeight`.
    /// Returns the warped JPEG's local file URL.
    ///
    /// Core Image's perspective-correction filter has its own
    /// coordinate convention — Y is flipped (origin at bottom-
    /// left). We flip the quad Y-coordinates before handing
    /// them in, so the result lands right-side-up.
    public static func warpAndSave(source: CGImage, quad: DetectedQuad) async -> URL? {
        let height = CGFloat(source.height)
        let ci = CIImage(cgImage: source)
        let filter = CIFilter(name: "CIPerspectiveCorrection")
        filter?.setValue(ci, forKey: kCIInputImageKey)
        filter?.setValue(CIVector(x: CGFloat(quad.tl.x), y: height - CGFloat(quad.tl.y)), forKey: "inputTopLeft")
        filter?.setValue(CIVector(x: CGFloat(quad.tr.x), y: height - CGFloat(quad.tr.y)), forKey: "inputTopRight")
        filter?.setValue(CIVector(x: CGFloat(quad.br.x), y: height - CGFloat(quad.br.y)), forKey: "inputBottomRight")
        filter?.setValue(CIVector(x: CGFloat(quad.bl.x), y: height - CGFloat(quad.bl.y)), forKey: "inputBottomLeft")
        guard let outputCi = filter?.outputImage else { return nil }
        // CIPerspectiveCorrection's output extent is the warped
        // rectangle's bounds — we want a fixed 1012×638 canvas,
        // so resample via a CIAffineTransform after the
        // perspective filter.
        let extent = outputCi.extent
        let sx = outputWidth  / extent.width
        let sy = outputHeight / extent.height
        let scaled = outputCi
            .transformed(by: CGAffineTransform(translationX: -extent.origin.x, y: -extent.origin.y))
            .transformed(by: CGAffineTransform(scaleX: sx, y: sy))

        let ctx = CIContext(options: nil)
        let outExtent = CGRect(x: 0, y: 0, width: outputWidth, height: outputHeight)
        guard let cg = ctx.createCGImage(scaled, from: outExtent) else { return nil }
        let warpedUiImage = UIImage(cgImage: cg)
        guard let jpegData = warpedUiImage.jpegData(compressionQuality: 0.92) else { return nil }
        // AttachmentStorage.write handles the
        // Application-Support directory creation + UUIDv7
        // filename + atomic write; it returns nil on any IO
        // failure which we pass straight through.
        return AttachmentStorage.write(jpegData, ext: "jpg")
    }
}
